Updated docker course development instructions (#2)
open-resources/opb_training: content/pl_docker/docker_course_dev.md

That's it!
You're ready to develop questions for the OPB!

> # Setting up local autograder container

Reviewer: This should be a Level Two heading (there can only be one level heading on a markdown page). I suggest `##` here, and then increase the indent level for subsequent sections and subsections by one - so `##` would go to `###`

Author: Resolved.

For development of questions that require an autograder, you will need to set up a local autograder container as well. 

## Important! (Windows Only)
If you are using Windows, the following commands will need to be executed in Windows Subsystem for Linux (WSL). You can do this two ways:
1. Open a command prompt and type `wsl` to open a WSL terminal
1. Look for the Windows Subsystem for Linux application in the Start Menu

> <img src="pl_images/wsl_icon_windows.png">

Reviewer: I don't fully understand here, what do you mean the "following commands" - do you mean "setting up your local Docker instance for auto grading will need to be done on WSL?"

Reviewer 2: Yes - on Windows for pl autograding to work you have to run docker through a wsl2 terminal instance, because of how docker works on windows (if you use the hyper-v version of windows docker backend it might be different, but its discouraged by docker to use that and it requires windows 10/11 pro (or enterprise maybe?) to use)

## Step 1: Create a local directory for the autograder

To run PrairieLearn locally with external grader and workspace support, create an empty directory to use to share job data between containers. This directory can live anywhere, but needs to be created first and referenced in the docker launch command. This directory only needs to be created once.
```bash
mkdir "$HOME/pl_ag_jobs"

> ```

Reviewer: add a line break between code-fence and preceding text for readability (apply everywhere)

## Step 2: Run the PrairieLearn Autograder container
Now, run PrairieLearn as usual, but with additional options. For example, if your course directory is in $HOME/pl-tam212 and the jobs directory created above is in $HOME/pl_ag_jobs, and you are using Linux or Mac OS X, the new command is as follows:
```bash

> docker run -it --rm -p 3000:3000 \
>     -v "$HOME/pl-tam212:/course" `# Replace the path with your course directory` \
>     -v "$HOME/pl_ag_jobs:/jobs" `# Map jobs directory into /jobs` \
>     -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
>     -v /var/run/docker.sock:/var/run/docker.sock `# Mount docker into itself so container can spawn others` \
>     prairielearn/prairielearn

Reviewer 2: ```suggestion
docker run -it --rm -p 3000:3000 \
    -v "$HOME/pl-opb-ind100:/course" `# Replace the path with your course directory` \
    -v "$HOME/pl_ag_jobs:/jobs" `# Map jobs directory into /jobs` \
    -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
    -v /var/run/docker.sock:/var/run/docker.sock `# Mount docker into itself so container can spawn others` \
    prairielearn/prairielearn
```

```

If you are on Windows, you can use the following command on the WSL 2 shell:
```bash

> docker run -it --rm -p 3000:3000 \
>     -v "$HOME/pl-tam212:/course" `# Replace the path with your course directory` \
>     -v "$HOME/pl_ag_jobs:/jobs" `# Map jobs directory into /jobs` \
>     -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
>     -v /var/run/docker.sock:/var/run/docker.sock `# Mount docker into itself so container can spawn others` \
>     --add-host=host.docker.internal:172.17.0.1 \
>     prairielearn/prairielearn

Reviewer 2: ```suggestion
docker run -it --rm -p 3000:3000 \
    -v "$HOME/pl-opb-ind100:/course" `# Replace the path with your course directory` \
    -v "$HOME/pl_ag_jobs:/jobs" `# Map jobs directory into /jobs` \
    -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
    -v /var/run/docker.sock:/var/run/docker.sock `# Mount docker into itself so container can spawn others` \
    --add-host=host.docker.internal:172.17.0.1 \
    prairielearn/prairielearn
```

> ```

Reviewer: This is great - thanks!

> *NOTE*: When replacing the path with your course directory on Windows, make sure to use the WSL path (e.g. `/mnt/c/Users/username/pl-tam212`) and __not__ the Windows path (e.g. `C:\Users\username\pl-tam212`).

Reviewer 2: ```suggestion
*NOTE*: When replacing the path with your course directory on Windows, make sure to use a POSIX-style path (e.g. `/mnt/c/Users/username/pl-opb-ind100`) and __not__ the Windows path (e.g. `C:\Users\username\pl-opb-ind100`).
```
Beyond the change here that I've put in the other comments, using the term POSIX-Style over WSL path just seems a bit more understandable since its the more common name for the path style

> ## Note: Grader images

Reviewer: Add a line break between the subheading and the content

That's it!
You're ready to develop questions for the OPB!

# Setting up local autograder container
For development of questions that require an autograder, you will need to set up a local autograder container as well. 

## Important! (Windows Only)
If you are using Windows, the following commands will need to be executed in Windows Subsystem for Linux (WSL). You can do this two ways:
1. Open a command prompt and type `wsl` to open a WSL terminal
1. Look for the Windows Subsystem for Linux application in the Start Menu
<img src="pl_images/wsl_icon_windows.png">

## Step 1: Pull the PrairieLearn Autograder container

Run the following command to pull the PrairieLearn Autograder image:

```bash

> docker pull prairielearn/grader-python

Reviewer 2: Should we mention things like the r-autograder (`prairielearn/grader-r`), since that could be used in the datasci or stats bank technically?

Author: We could. Though I actually found that if the container isn't present, PL auto pulls the container. So step 1 technically isn't even necessary for the autograder to run. As long as the Docker command is properly configured, it should work.

Reviewer 2: Maybe its better to just exclude the step then if prairielearn pulls the containers, and just make a note somewhere that first runs may take longer since prairielearn/docker will pull the needed docker containers if they don't exist yet - keeps the guide simpler that way I think

Author: I was thinking that as I was making the document. I made some changes. How does this look?

Reviewer 2: Looks good to me!
